docs/adr/0002-recipe-rewrite.md

```diff
@@ -0,0 +1,254 @@
+
+`RecipeInputItem`s describe a single slot of a recipe and determines what
+items match it. There can be a single item that matches (see `RecipeInputSlimefunItem`,
+`RecipeInputItemStack`), or a list (tag) of items all of which can be used
+in that slot (see `RecipeInputGroup`, `RecipeInputTag`).
+
+`RecipeOutput`s are just a list of `RecipeOutputItem`s, all of which are crafted by the recipe.
+
+An `RecipeOutputItem`s controls how an output is generated when the recipe is
+crafted. It can be a single item (see `RecipeOutputItemStack`, `RecipeOutputSlimefunItem`),
+or a group of items each with a certain weight of being output (see `RecipeOutputGroup`).
+
+#### Examples (pseudocode)
+
+Here are the inputs and outputs of the recipe for a vanilla torch
+
+```txt
+RecipeInput (
+    {
+        EMPTY, EMPTY, EMPTY
+        EMPTY, RecipeInputGroup(COAL, CHARCOAL), EMPTY,
+        EMPTY, RecipeInputItemStack(STICK), EMPTY
+    },
+    SHAPED
+)
+RecipeOutput (
+    RecipeOutputItemStack(4 x TORCH)
+)
+```
+
+Here are the inputs and outputs of a gold pan
+
+```txt
+RecipeInput (
+    { RecipeOutputItemStack(GRAVEL) },
+    SUBSET
+)
+RecipeOutput (
+    RecipeOutputGroup(
+        40 RecipeOutputItemStack(FLINT)
+        5  RecipeOutputItemStack(IRON_NUGGET)
+        20 RecipeOutputItemStack(CLAY_BALL)
+        35 RecipeOutputSlimefunItem(SIFTED_ORE)
+    )
+)
+```
+
+This would remove the need to use ItemSettings to determine the gold pan weights
+
+### RecipeService
+
+This is the public interface for the recipe system, there are methods here to add,
+load, save, and search recipes. It also stores a map of `MatchProcedures` and
+`RecipeType` by key for conversions from a string
+
+### JSON Serialization
+
+All recipes are able to be serialized to and deserialized
+from JSON. The schemas are shown below.
+
+Here, `key` is the string representation of a namespaced key
+
+`Recipe`
+
+```txt
+{
+    "input"?: RecipeInput
+    "output"?: RecipeOutput
+    "type": key | key[]
+    "energy"?: int
+    "crafting-time"?: int
+    "permission-node"?: string | string[]
+}
+```
+
+The recipe deserializer technically needs a `__filename` field, but it is
+inserted when the file is read, so it isn't (and shouldn't) be in the schema
+
+`RecipeInput`
+
+```txt
+{
+    "items": string | string[]
+    "key": {
+        [key: string]: RecipeInputItem
+    }
+    "match"?: key
+}
+```
+
+`RecipeOutput`
+
+```txt
+{
+    "items": RecipeOutputItem[]
+}
+```
+
+`RecipeInputItem`*
+
+```txt
+{
+    "id": key
+    "amount"?: int
+    "durability"?: int
+} | {
+    "tag": key
+    "amount"?: int
+    "durability"?: int
+} | {
+    "group": RecipeInputItem[]
+}
+```
+
+`RecipeOutputItem`*
+
+```txt
+{
+    "id": key
+    "amount"?: int
+} | {
+    "group": RecipeInputItem[]
+    "weights"?: int[]
+}
+```
+
+*In addition to those schemas, items can be in short form:
+
+- Single items: `<namespace>:<id>|<amount>`
+- Tags: `#<namespace>:<id>|<amount>`
```

Comment:
tbh i'm not a big fan of this, i'd rather be more explicit. I can feel this getting unwieldy in the future

Reply:
just to be clear, the short form is only used for items that are *just* an id (mc or sf) and amount

i think this would just be a lot cleaner when writing recipes, since most inputs won't have any other fields anyways

for example
```
{
    "id": "ELECTROMAGNET",
    "input": {
        "items": [
            "123",
            " 4 ",
            "   "
        ],
        "key": {
            "1": "slimefun:nickel_ingot",
            "2": "slimefun:magnet",
            "3": "slimefun:cobalt_ingot",
            "4": "slimefun:battery"
        },
        "match": "slimefun:shaped"
    },
    "output": {
        "items": [
            "slimefun:electromagnet"
        ]
    },
    "type": "slimefun:enhanced_crafting_table"
}
```
vs
```
{
    "id": "ELECTROMAGNET",
    "input": {
        "items": [
            "123",
            " 4 ",
            "   "
        ],
        "key": {
            "1": {
                "id": "slimefun:nickel_ingot"
            },
            "2": {
                "id": "slimefun:magnet"
            },
            "3": {
                "id": "slimefun:cobalt_ingot"
            },
            "4": {
                "id": "slimefun:battery"
            }
        },
        "match": "slimefun:shaped"
    },
    "output": {
        "items": [
            {
                "id": "slimefun:electromagnet"
            }
        ]
    },
    "type": "slimefun:enhanced_crafting_table"
}
```

i forgot to mention the `|<amount>` bit was optional if its just 1; ill edit that in.